src/binary_deserializer.ml

```diff
@@ -0,0 +1,359 @@
+(* binary format specification:
+   https://webassembly.github.io/spec/core/binary/modules.html#binary-importsec *)
+
+open Simplified
+open Syntax
+open Types
+open Binary_basic
+
+type ('a, 'b) import =
+  | Func of int
+  | Table of limits * 'a ref_type
+  | Mem of limits
+  | Global of mut * 'b val_type
+
+let consume_to_end x input = (x, Input.sub ~pos:0 ~len:0 input)
+
+let magic_check bytes =
+  let magic = Bytes.sub_string bytes 0 4 in
+  match String.equal magic "\x00\x61\x73\x6d" with
+  | true -> Ok ()
+  | false -> assert false
+
+let version_check bytes =
+  let version = Bytes.sub_string bytes 4 4 in
+  match String.equal version "\x01\x00\x00\x00" with
+  | true -> Ok ()
+  | false -> assert false
+
+let section_parse input ~expected_id default section_content_parse =
+  match Input.get0 input with
+  | Some id when id = expected_id ->
+    let input = Input.sub_suffix 1 input in
+    let size, input = read_U32 input in
+    let section_input = Input.sub_prefix size input in
+    let next_input = Input.sub_suffix size input in
+    let res, after_section_input = section_content_parse section_input in
+    assert (Input.is_empty after_section_input);
+    (res, next_input)
+  | _ -> (default, input)
+
+let vector parse_elt input =
+  let nb_elt, input = read_U32 input in
+  let rec loop loop_id input acc =
+    if nb_elt = loop_id then (List.rev acc, input)
+    else
+      let acc_elt, input = parse_elt loop_id input in
+      let acc = acc_elt :: acc in
+      loop (loop_id + 1) input acc
+  in
+  loop 0 input []
+
+let section_custom input =
+  print_endline "custom";
+  section_parse input ~expected_id:'\x00' () @@ consume_to_end ()
+
+let section_type input =
+  print_endline "type";
+  section_parse input ~expected_id:'\x01' []
+    (vector (fun id input ->
+         let fcttype, input = read_byte input in
+         assert (fcttype = '\x60');
+         let params, input =
+           vector (fun _id input -> deserialize_valtype input) input
+         in
+         let results, input =
+           vector (fun _id input -> deserialize_valtype input) input
+         in
+         let params = List.map (fun param -> (None, param)) params in
+         (Bt_raw (Some (Raw id), (params, results)), input) ) )
+
+let section_import input =
+  print_endline "import";
+  section_parse input ~expected_id:'\x02' []
+  @@ vector (fun _id input ->
+         let modul, input = vector (fun _id input -> read_byte input) input in
+         let name, input = vector (fun _id input -> read_byte input) input in
+         let modul = String.of_seq (List.to_seq modul) in
```

Comment:
A function `string_of_char_list` would be nice to have defined separately (it can also probably be implemented in a better way ;))

Reply:
I don't know if this implementation would be better (in terms of complexity)  :
```ocaml
let string_of_char_list char_list =
  (* String.of_seq (List.to_seq char_list) *)
  let char_arr = Array.of_list char_list in
  String.init (Array.length char_arr) (fun i -> Array.get char_arr i)
```

Reply:
Something like:

```ocaml
let buf = Buffer.create 64 in
List.iter (Buffer.add_char buf) char_list;
Buffer.contents buf
```

would be better. :)

Reply:
OCaml Buffer module :ok_hand: